Make Native Image opt-out

hubertp opened the pull request:
### Pull Request Description

This change makes Native Image build opt-out rather than opt-in. To use JVM mode, IDE has to start with `--jvm` or `ENSO_LAUNCHER=shell`.

An alternative approach to https://github.com/enso-org/enso/pull/12501. Closes #12483.
Likely also addresses https://github.com/enso-org/enso/issues/12303.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Make Native Image opt-out

This change makes Native Image build opt-out rather than opt-in.
To use JVM mode, IDE has to start with `ENSO_LAUNCHER=shell`.

JaroslavTulach commented on build_tools/build/src/engine/context.rs:
```diff
@@ -281,6 +281,7 @@ impl RunContext {
         // we don't want to call this in environments like GH-hosted runners.
 
         // === Build project-manager distribution and native image ===
+        env::ENSO_LAUNCHER.set(&engine::EngineLauncher::Native)?;
```
What is the need for this change? 
- As far as I know everything has been already built properly
- _native mode_ is supposed to be the default
- there should be no need to change anything in build system
- the only change needed is to use _native image_ `enso` launcher in `project-manager`

hubertp replied:
That's the change that enables NI when building IDE, as needed also in [your PR](https://github.com/enso-org/enso/pull/12501#issuecomment-2728235731)

JaroslavTulach replied:
Is the change needed? 


When I download [regular app image](https://productionresultssa10.blob.core.windows.net/actions-results/0b18635e-d118-4346-a018-aad6ffb7993a/workflow-job-run-c21fb4be-a48c-57e7-6e26-d8556d9a7861/artifacts/1b948be97a1cf5a2515f9e4107a22a7c21b8386139dfd6d3d4ca618253d073f8.zip?rscd=attachment%3B+filename%3D%22ide-linux-amd64.zip%22&se=2025-03-17T09%3A09%3A20Z&sig=lAU6IbqK5btWbR8VkYnauuDm0Eha1tVSSReFA57jomk%3D&ske=2025-03-17T18%3A40%3A11Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-03-17T06%3A40%3A11Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-01-05&sp=r&spr=https&sr=b&st=2025-03-17T08%3A59%3A15Z&sv=2025-01-05) built [without this change](https://github.com/enso-org/enso/actions/runs/13892458438) I see:
```
$ ./enso-linux-x86_64-2025.1.1-nightly.2025.3.17.1.AppImage --appimage-extract
$ ls -l squashfs-root/resources/enso/dist/2025.1.1-nightly.2025.3.17.1/bin/enso -h
-rwxr-xr-x 1 devel devel 480M bře 17 11:30 squashfs-root/resources/enso/dist/2025.1.1-nightly.2025.3.17.1/bin/enso
```
e.g. the `enso` _native image_ binary is present in the `AppImage`.

hubertp replied:
It appears so, yes. We only build in NI in release mode [it seems](https://github.com/enso-org/enso/blob/develop/project/GraalVM.scala#L21).

JaroslavTulach replied:
> We only build in NI in release mode [it seems](https://github.com/enso-org/enso/blob/develop/project/GraalVM.scala#L21).

- that's correct. NI shall be built only in release mode
- as #12296 puts it: 
   - during the _development_ we continue to use `shell` launchers by default
   - follow instructions in #12117 to develop within _native mode_

>> Is the change needed?
> It appears so, yes.

Why should this change be needed? 
- `AppImage` at https://github.com/enso-org/enso/actions/runs/13892458438 contains `enso` binary already
- that's because the `AppImage` produced by [Nightly Release CI](https://github.com/enso-org/enso/actions/workflows/nightly.yml) is **built in release mode**
- e.g. [EnsoLauncher.native](https://github.com/enso-org/enso/blob/develop/project/GraalVM.scala#L67) is already `true` even without `ENSO_LAUNCHER` being set

Is there an artifact improperly built with _"shell launchers"_?
- I am not aware of it
- Are you? Do you have a link?
- If not, then there is no point setting `env::ENSO_LAUNCHER.set(&engine::EngineLauncher::Native)?` environment variables in CI
- the whole point of #12117 was to avoid `ENSO_LAUNCHER` manipulation in the CI

hubertp replied:
That's not going catch dev mistakes when they push changes that require updates to reflection config.

hubertp replied:
One way to find out how big of an issue that will be. This can be improved on.

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Runner.scala:
```diff
@@ -224,18 +224,28 @@ class Runner(
-                  JavaExecCommand.forRuntime(runtime)
-                )
+            val ensoLauncher      = Option(System.getenv(Runner.LAUNCHER_ENV_NAME))
+            val requiresJVMRunner = ensoLauncher.exists(_.equals("shell"))
```
The name `shell` is technically misleading. There is no _shell_ anywhere 
- in fact `project-manager` bypasses the launcher completely
- invokes directly `java`command with arguments referring to various JARs inside of engine
- that needlessly enlarges the API between `project-manager` and engine
- ideally `project-manager` should only add `--jvm` and continue to invoke the `enso` launcher

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Runner.scala:
```diff
@@ -303,3 +313,9 @@ class Runner(
+
+object Runner {
+
+  private val LAUNCHER_ENV_NAME = "ENSO_LAUNCHER"
```
Compared to [2024.5.2 release](https://github.com/enso-org/enso/releases/tag/2024.5.2) there is new, user facing environment variable `ENSO_LAUNCHER` - e.g. new API. Should such environment variable be supported in the next official release, it has to be documented.

JaroslavTulach replied:
In the most recent version of this PR, the `ENSO_LAUNCHER` environment variable is gone. Good.

JaroslavTulach replied:
- #13225 attempts to
- remove support for `ENSO_LAUNCHER=shell`

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Runner.scala:
```diff
@@ -224,18 +224,28 @@ class Runner(
+                logger
+              ) match {
+                case Some(cmd) =>
+                  prepareAndRunCommand(engine, cmd)
```
As [discovered recently](https://github.com/enso-org/enso/pull/12501#issuecomment-2726213993):

> `ENSO_JVM_OPTS` are used as enso **executable arguments** when launching _native image_ `enso` executable. That means ...  the `ENSO_JVM_OPTS` **change their interpretation** when changing JVM/NI mode.

- what is the meaning of `ENSO_JVM_OPTS` in the proposed `ENSO_LAUNCHER=shell` mode?
- should `ENSO_LAUNCHER` be exposed to end users in next version
- the proper meaning of `ENSO_JVM_OPTS` in non-JVM mode has to be documented

JaroslavTulach replied:
OK. Let's have official `--jvm` to switch to the JVM mode. Let's leave the (less prominent) `ENSO_JVM_OPTS` variable to influence options to both NI and JVM modes.

JaroslavTulach reviewed (commented):
- `ENSO_LAUNCHER` would be a new user facing API
- it has to be documented
- meaning of `ENSO_JVM_OPTS` and `ENSO_LAUNCHER` combination is not clear
- such a combination has to be documented/tested should `ENSO_LAUNCHER` environment variable really become support API for the next release

JaroslavTulach commented:
According to @jdunkerley [some users are using `ENSO_JVM_OPTS`](https://github.com/enso-org/enso/pull/12501#issuecomment-2727169765) already:
- this PR would **break their workflows**
    - very likely the `project-manager` would not be able to invoke the engine at all
    - failing with uncomprehensible failure
- it changes the meaning of `ENSO_JVM_OPTS`
- contary to #12501 which 
   - **attempts to keep compatiblity** 
   - keeps meaning of `ENSO_JVM_OPTS` unchanged

hubertp commented:
I agree that `ENSO_LAUNCHER` is not the best way. Since the other PR got no-merge, I will update this PR so that `project-manager`/IDE accepts `--jvm` instead. That should be fairly non-controversial and easy to add.

JaroslavTulach commented:
As the one who [argued in favor of --jvm](https://github.com/enso-org/enso/pull/11880#pullrequestreview-2528578132) in the past, I shall support the `project-manager --jvm` approach....

> I will update this PR so that `project-manager`/IDE accepts `--jvm` instead. That should be fairly non-controversial and easy to add.

...and I will!
- `project-manager --jvm` option is consistent with `enso --jvm`
- we wanted such a consistency and we still want it - e.g. it seems like a good solution
- `--jvm` focuses on the core issues of today: _opt-out from native image mode_
- we leave the memory settings & co. options for power users via `ENSO_JVM_OPTS` for now

One question:
- how does one pass `--jvm` option to `AppImage`?

hubertp committed:
Pass around `--jvm` flag in PM/LS

hubertp commented:
> * how does one pass `--jvm` option to `AppImage`?

I'm already on it.

hubertp committed:
Allow for `--jvm` flag in AppImage

`--jvm` is passed to project manager and enables JVM mode.

JaroslavTulach commented on app/ide-desktop/client/src/index.ts:
```diff
@@ -349,7 +349,9 @@ class App {
         : []
-      const backendOpts = [...backendVerboseOpts, ...backendProfileOpts]
+      const backendJvmOpts =
+        this.args.options.jvm.value ? ['--jvm'] : []
```
77178f991b65ad7c1668f9cbabf40772c3f3b862 seems simple. If it works, then great. Let's get this in.

hubertp committed:
changelog update

hubertp committed:
linter

JaroslavTulach reviewed (commented):
- I like how the PR is shaping with `--jvm` argument
- I still believe [context.rs change](https://github.com/enso-org/enso/pull/12515/files#r1998605116) shouldn't be part of this PR

hubertp committed:
Unbreak build

One Google_Api_Test is known to be outdated/broken. Increased timeout
for building index - Windows appears to have problems.

hubertp committed:
nits

hubertp committed:
Disable building NI for non-releases

As requested.

hubertp commented:
Ready for review, folks.

JaroslavTulach commented on docs/infrastructure/native-image.md:
```diff
@@ -200,9 +200,10 @@ safely.
-The Native Image generation for the Engine Runner is currently in a preview
-state. It is triggered by `ENSO_LAUNCHER` environment variable. Its value can be
-one of the following:
+The Native Image generation for the Engine Runner is on my default for releases.
```
```suggestion
The Native Image generation for the Engine Runner is on by default for releases.
```

JaroslavTulach reviewed (approved):
With 05329f9309fc326808bfc4569bec2c358c51ff76 it all seems fine to me.

somebody1234 reviewed (approved):
approving changes in `app/ide-desktop`

hubertp committed:
fix tests

hubertp committed:
Update docs/infrastructure/native-image.md

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

4e6 commented on project/DistributionPackage.scala:
```diff
@@ -276,7 +276,7 @@ object DistributionPackage {
           ).run
           // Poor man's solution to stuck index generation
-          val GENERATING_INDEX_TIMEOUT = 60 * 2 // 2 minutes
+          val GENERATING_INDEX_TIMEOUT = 60 * 4 // 2 minutes
```
Forgot to cleanup or
```suggestion
          val GENERATING_INDEX_TIMEOUT = 60 * 4 // 4 minutes
```

4e6 replied:
in ideal world
```scala
scala> {
     | import scala.concurrent.duration._
     | 4.minutes.toSeconds
     | }
val res4: Long = 240
```

hubertp replied:
Good catch, I will add it to next PR. Didn't want to trigger yet another build, it would take ages and this needed to get it to nightly.

4e6 commented on test/Google_Api_Test/src/Main.enso:
```diff
@@ -10,7 +10,7 @@ main filter=Nothing =
 
 add_specs suite_builder =
-    suite_builder.group "Google Spreadsheets" group_builder->
+    suite_builder.group "Google Spreadsheets" pending="outdated" group_builder->
```
seems unrelated

hubertp replied:
Related. NI would potentially crash on this test and we know it's broken.

JaroslavTulach commented on engine/launcher/src/main/scala/org/enso/launcher/cli/LauncherApplication.scala:
```diff
@@ -109,6 +112,12 @@ object LauncherApplication {
       "jvm",
       "These parameters will be passed to the launched JVM as -DKEY=VALUE."
     )
+  private def jvmMode =
```
- The `jvmMode` flag introduced by this PR is causing some troubles in #13225
- why do we have this flag at all?
- is it needed for something important?
- something that is tested?
- can we [delete the jvmMode flag](https://github.com/enso-org/enso/pull/13225#discussion_r2134403383)?
